server/platform/controllers/action_types.go

```diff
@@ -0,0 +1,181 @@
+			OrganizationID: a.OrganizationID,
+			Name:           a.Name,
+			Version:        a.Version,
+			Properties:     a.Properties,
+			Archived:       a.Archived,
+			CreatedAt:      a.CreatedAt,
+		})
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func (ac ActionTypesController) Show(c *gin.Context) {
+	user := c.MustGet("user").(*db.User)
+
+	id := uuid.FromStringOrNil(c.Param("id"))
+	action, err := ac.App.DB.GetActionTypeByUUID(c.Request.Context(), user.OrganizationID, id)
+	if err != nil {
+		ac.App.Logger.Warn("actiontypes.show.getActiontype", zap.Error(err))
+		c.AbortWithStatus(http.StatusNotFound)
+		return
+	}
+
+	res := responses.Action{
+		ID:             action.ID,
+		OrganizationID: action.OrganizationID,
+		Name:           action.Name,
+		Version:        action.Version,
+		Properties:     action.Properties,
+		CreatedAt:      action.CreatedAt,
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func (ac ActionTypesController) Create(c *gin.Context) {
+	var req requests.ActionTypeCreate
+	if err := c.ShouldBindJSON(&req); err != nil {
+		ac.App.Logger.Warn("actiontypes.create.parseRequest", zap.Error(err))
+		c.AbortWithStatus(http.StatusBadRequest)
+		return
+	}
+
+	user := c.MustGet("user").(*db.User)
+
+	properties, _ := json.Marshal(req.Properties)
+	name := req.Name
+	version := req.Version
+
+	archivedTypes, err := ac.App.DB.GetActionTypesByName(c, user.OrganizationID, name)
+	if err != nil {
+		ac.App.Logger.Error("actiontypes.create.listArchivedVersions", zap.Error(err))
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	for _, archivedType := range archivedTypes {
+		ac.App.DB.UnArchiveActionType(c, user.OrganizationID, archivedType.ID)
+		version = archivedType.Version + 1
+	}
+
+	args := db.CreateActionTypeParams{
+		OrganizationID: user.OrganizationID,
+		Name:           req.Name,
+		Version:        version,
+		Properties:     properties,
+	}
+
+	actionType, err := ac.App.DB.CreateActionType(c.Request.Context(), args)
+	if err != nil {
+		ac.App.Logger.Error("actiontypes.create.createActiontype", zap.Error(err))
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	res := responses.ActionType{
+		ID:             actionType.ID,
+		OrganizationID: actionType.OrganizationID,
+		Name:           actionType.Name,
+		Version:        actionType.Version,
+		Properties:     actionType.Properties,
+		CreatedAt:      actionType.CreatedAt,
+	}
+
+	c.JSON(http.StatusOK, res)
+}
+
+func (ac ActionTypesController) Archive(c *gin.Context) {
+	user := c.MustGet("user").(*db.User)
+
+	name := c.Param("id")
+	if err := ac.App.DB.ArchiveActionType(c.Request.Context(), user.OrganizationID, name); err != nil {
+		ac.App.Logger.Error("actiontypes.archive.archiveType", zap.Error(err))
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	c.AbortWithStatus(http.StatusNoContent)
+}
+
+type Usage struct {
+	InUse bool `json:"inUse"`
+}
+
+func (ac ActionTypesController) InUse(c *gin.Context) {
```

Comment:
Would it make sense to add a boolean flag to the list of action types with the information if it is already in use. You could achieve this by joining the `actions` table on querying the `action_types`.
This way we can skip the whole `InUse` endpoint.

Reply:
How could this be achieved?
INNER JOIN with action_type.id = actions.type_id
can i then check if there is no join with the actions table?

Reply:
You could use a LEFT JOIN and check if the joined row is NULL for the ‘action.id column‘.